### PR TITLE
Placeholder painting

### PR DESCRIPTION
**Pull request by spireken**

### CODE REVIEW

File: Applications/Spire/Include/Spire/Ui/TextBox.hpp [hunk 1]
```diff
@@ -126,25 +126,26 @@ namespace Styles {
...
       std::shared_ptr<TextModel> m_model;
-      Box* m_box;
+      PlaceholderBox* m_box;
       QLineEdit* m_line_edit;
       StyleProperties m_line_edit_styles;
       boost::signals2::scoped_connection m_style_connection;
       boost::signals2::scoped_connection m_placeholder_style_connection;
-      QLabel* m_placeholder;
       StyleProperties m_placeholder_styles;
+      QMargins m_placeholder_margins;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/TextBox.hpp:146]: Can you encapsulate the placeholder further? Things like ```m_placeholder_style_connection```, ```m_placeholder_styles```, ```commit_placeholder_style```, do as much encapsulation as possible (single responsibility principle).

**Review by kranar — COMMENTED** [review]
This is so much nicer than before.

File: Applications/Spire/Source/Ui/TextBox.cpp [hunk 2]
```diff
@@ -96,6 +97,75 @@ struct TextBox::TextValidator : QValidator {
...
+      m_text_color = properties.m_text_color;
+      m_margins = margins;
+      update();
+    }
+
+  protected:
+    void paintEvent(QPaintEvent* event) {
+      Box::paintEvent(event);
+      if(m_is_text_visible) {
+        auto painter = QPainter(this);
```
[the remark this answers is not in the document]
Comment (spireken) [reply Applications/Spire/Source/Ui/TextBox.cpp:142]: I can try, but there's a warning in the [documentation](https://doc.qt.io/qt-5/qpainter.html#details):

"Warning: When the paintdevice is a widget, QPainter can only be used inside a paintEvent() function or in a function called by paintEvent()."

I'm not sure if 'used' means calling begin() and end() to begin and end a painting operation, or if used means constructing a QPainter outside of the paintEvent at all. Regardless, I'll investigate.

File: Applications/Spire/Source/Ui/TextBox.cpp [hunk 3]
```diff
@@ -135,15 +206,7 @@ TextBox::TextBox(std::shared_ptr<TextModel> model, QWidget* parent)
...
-  m_placeholder = new QLabel();
-  m_placeholder->setCursor(m_line_edit->cursor());
-  m_placeholder->setTextFormat(Qt::PlainText);
-  m_placeholder->setMargin(0);
-  m_placeholder->setIndent(0);
-  m_placeholder->setTextInteractionFlags(Qt::NoTextInteraction);
-  m_placeholder->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
-  layers->add(m_placeholder);
-  m_box = new Box(layers);
+  m_box = new PlaceholderBox(layers);
```
Comment (shaobinspire) [inline Applications/Spire/Source/Ui/TextBox.cpp:209]: I think 'layers' is no longer needed.

File: Applications/Spire/Source/Ui/TextBox.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (spireken) [reply Applications/Spire/Source/Ui/TextBox.cpp:142]: At least in regards to painting onto a QWidget, calling setFont and setPen on a QPainter requires a valid QPaintEngine, which is only valid for a widget when in the paintEvent method. Qt has special cases to handle a QPixmap and QImage which is why painting to them works outside of paintEvent.

**Comment by kranar** [comment]
Then thats the end of that idea.

**Pull request merged**